orchestra/actions/util/__init__.py

```diff
@@ -0,0 +1,126 @@
+    :param script: the script to run
+    :param environment: optional additional environment variables
+    :param decode_as: decode the script output using this encoding
+    :return: the stdout produced by the script
+    """
+    return _get_script_output(script, environment=environment, check_returncode=True, decode_as=decode_as)
+
+
+def try_get_script_output(script, environment: OrderedDict = None, decode_as="utf-8"):
+    """Helper for getting stdout of a script. If the script returns nonzero an empty output will be returned.
+    :param script: the script to run
+    :param environment: optional additional environment variables
+    :param decode_as: decode the script output using this encoding
+    :return: the stdout produced by the script
+    """
+    return _get_script_output(script, environment=environment, check_returncode=False, decode_as=decode_as)
+
+
+def run_internal_subprocess(
+        argv,
+        environment: [OrderedDict, dict] = None,
+        cwd=None,
+):
+    """Helper for running an internal subprocess.
+    If the subprocess returns a nonzero exit code an OrchestraException is raised.
+    :param argv: the argv passed to subprocess.run
+    :param environment: environment variables
+    :param cwd: if not None, the command is executed in the specified path
+    :return: a subprocess.CompletedProcess instance
+    """
+    return _run_internal_subprocess(argv, environment=environment, cwd=cwd, check_returncode=True)
+
+
+def try_run_internal_subprocess(
+        argv,
+        environment: [OrderedDict, dict] = None,
+        cwd=None,
+):
+    """Helper for running an internal subprocess.
+    :param argv: the argv passed to subprocess.run
+    :param environment: environment variables
+    :param cwd: if not None, the command is executed in the specified path
+    :return: a subprocess.CompletedProcess instance
+    """
+    return _run_internal_subprocess(argv, environment=environment, cwd=cwd, check_returncode=False)
+
+
+def get_subprocess_output(
+        argv,
+        environment=None,
+        decode_as="utf-8",
+):
+    """
+    Helper to run a subprocess and get its output.
+    If the subprocess returns a nonzero exit code an OrchestraException is raised.
+    :param argv: the argv passed to subprocess.run
+    :param environment: environment variables
+    :param decode_as: decode the output using this encoding
+    :return: the decoded stdout of the subprocess
+    """
+    return _get_subprocess_output(argv, environment=environment, decode_as=decode_as, check_returncode=True)
+
+
+def try_get_subprocess_output(
+        argv,
+        environment=None,
+        decode_as="utf-8",
+):
+    """
+    Helper to run a subprocess and get its output.
+    :param argv: the argv passed to subprocess.run
+    :param environment: environment variables
+    :param decode_as: decode the output using this encoding
+    :return: the decoded stdout of the subprocess
```

Comment:
Can you please document what's the return value is in case of failure? `None`, I guess?

Reply:
Good point. As of now whatever the program managed to put on stdout gets returned (i.e. possibly an empty string or partial output).
I understand this means it might not be easy/possible to recognize if the subprocess failed with this helper.
If that's needed one can use `get_subprocess_output` which raises an exception if the subprocess failed, or the lower level `try_run_internal_subprocess` which will return a `subprocess.CompletedProcess` which allows to access both the return code and the stdout.

Reply:
Mmmh, a `try_...` method should either succeed or clearly report failure.
I suggest to return `None` in case of failure, or at least return a tuple `stdout, success`, if you think it's more appropriate.
As of now it's quite volatile.

Reply:
Hmm, since the `try_*` functions are not very used at this point it might be better to just remove them and ask the users to use a `try ... catch` block if the subprocess called is expected to fail.
What do you think?

Reply:
If it was pervasively used I'd object, but given it's used only for `git ls-remote` it might make sense.
It's more verbose, but more explicit. OKGO.

Reply:
I rebased on current `master` and removed the  `try_*` helpers